CONTRIBUTING.md

```diff
@@ -474,3 +482,126 @@ that should be reported, so there's no need to make these messages readable to t
+
+
+### Auth overview
+
+When running on local computer, you should authenticate as user and not pet service account (SA). This way, you don't need to export pet SA keys, which is a security risk.
+
+<table>
+<tr>
+<td></td>
+<td>
+
+Regular command (eg `terra workspace`)
+</td>
+<td>Tool command (Docker and local process mode)</td>
+</tr>
+<tr>
+<td>Local computer</td>
+<td valign="top">
+
+**User Terra OAuth**
+
+`terra auth login` CLI goes through oauth flow and saves credentials in `.terra/StoredCredential`. No `gcloud` involved. Requests to services such as WorkspaceManager use an access token obtained from `.terra/StoredCredential`.
+</td>
+<td>
+
+**gcloud user credentials or user ADC, depending on tool**
+
+Most tools accept both (user credentials or ADC). Even though we don't need ADC for most tools, [we currently require ADC for all tools](https://github.com/DataBiosphere/terra-cli/blob/79a938e56f2d95111aeec54175b6d38d9e5deb79/src/main/java/bio/terra/cli/app/DockerCommandRunner.java#L98).
+
+Normally ADC is used with SA. When running Terra on a local
+computer, you are using ADC with your user credentials. This is unusual. (And as mentioned earlier,
+this is because we currently require ADC for all tools.) Please ignore this warning:
+`WARNING: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead`.
+
+[Nextflow only works with ADC.](https://www.nextflow.io/docs/latest/google.html#credentials) User is asked to run `gcloud auth application-default login`, which writes  `.config/gcloud/application_default_credentials.json`.
+
+In Docker mode, [we mount `.config/gcloud`](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/main/java/bio/terra/cli/app/DockerCommandRunner.java#L80).
+
+</td>
+</tr>
+<tr>
+<td>GCP notebook</td>
+<td>
+
+**Pet ADC** (if you use `--use-app-default-credentials`)
+
+Notebooks can run terra CLI with  [`--use-app-default-credentials`](https://github.com/DataBiosphere/terra-cli/pull/197/files).
+terra will pick up creds from ADC instead of OAuth (which involves interacting with a browser).
+</td>
+<td valign="top">
+
+**Pet ADC**
+
+ADC is automatically configured for GCE VMs.
+</td>
+</tr>
+<tr>
+<td>Unit test</td>
+<td valign="top">
+
+**Test user Terra OAuth**
+
+At the beginning of each test, before running terra CLI, [test user is logged in](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/test/java/harness/baseclasses/SingleWorkspaceUnit.java#L30).
+
+`TestUser.login()` [writes `.terra/StoredCredential`](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/test/java/harness/TestUser.java#L76), using domain-wide delegation to avoid browser flow.
+</td>
+<td>
+
+**Test user**
+
+Auth is complicated because we are not using Pet SA key (which is a security risk).
```

Comment:
You already describe above about not using Pet SA.

Maybe the caption here should be something along the lines of "Test authenticate using an end-user test account"?

Reply:
I took out `(which is a security risk)`. I left in `Auth is complicated` part. The reason we have to do bespoke, complicated things for each tool is because we're not using SA key.

If there's a new tool, or if any these break for existing tools, it's non-trivial to figure out how to authenticate. It's not documented anywhere. I had to use pdb debugger to figure out what to do.